### PR TITLE
Optionally expose time to screen shaders

### PR DESCRIPTION
**Pull request by scorpion-26**
This PR adds a way to have access to time in screen shaders. Time is essential for advanced, animated post-process shader effects (i.e. CRT effect, glitch effects, ...). 
Since that requires damage tracking to be disabled, I implemented it as "Pay for what you use". This means time is only added, if it is actually used in the shader; and users that don't use it have no disadvantage. 
If damage tracking is enabled and time is used an error is displayed. 

Implements part of https://github.com/hyprwm/Hyprland/issues/1502

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
This requires, that "damage_tracking = 0" updates every frame, which seems to be the case.

#### Is it ready for merging, or does it need work?
Ready. Criticisms/suggestions welcome though.

### CODE REVIEW

File: src/render/OpenGL.hpp [hunk 1]
```diff
@@ -133,6 +134,7 @@ class CHyprOpenGLImpl {
     bool              m_bApplyFinalShader = false;
 
     CShader           m_sFinalScreenShader;
+    CTimer            m_Timer;
```
Comment (vaxerski) [inline src/render/OpenGL.hpp:137]: terrible name. I propose `m_tGlobalTime` or sumn

File: src/render/OpenGL.cpp [hunk 2]
```diff
@@ -511,6 +520,14 @@ void CHyprOpenGLImpl::renderTextureInternalWithDamage(const CTexture& tex, wlr_b
     glUniformMatrix3fv(shader->proj, 1, GL_FALSE, glMatrix);
 #endif
     glUniform1i(shader->tex, 0);
+
+    if (usingFinalShader && g_pConfigManager->getInt("debug:damage_tracking") == 0) {
+        glUniform1f(shader->time, m_Timer.getSeconds());
```
Comment (vaxerski) [inline src/render/OpenGL.cpp:525]: this is not an issue per se - but wouldn't milliseconds with an int make more sense here? Dunno if opengl uses 64 or 32 bit ints though - if 32 then it's not really a good solution imho (overflow after 596h) but if it's 64-bit we can make it have zero inaccuracies.

**Review by vaxerski — CHANGES_REQUESTED** [review]
One comment and one requested change. In general, this looks reasonable to me.

File: src/render/OpenGL.hpp (hunk 1 above)
Comment (scorpion-26) [reply src/render/OpenGL.hpp]: Fixed

File: src/render/OpenGL.cpp (hunk 2 above)
Comment (scorpion-26) [reply src/render/OpenGL.cpp]: Good question. In terms of shader programming, ```float time;``` in seconds would be way less awkward than ```int time;``` in milliseconds and definitely more common. We're limited anyways by the return value of the internal call to ```std::...::duration::count()``` (which is a 32bit long and in milliseconds for ```getSeconds```). ```GLint``` is also typedef'ed to ```int``` on the C++ side.

Comment (vaxerski) [reply src/render/OpenGL.cpp]: > We're limited anyways by the return value of the internal call to `std::...::duration::count()` (which is a 32bit long and in milliseconds for getSeconds).

umm...
image

> `GLint` is also typedef'ed to `int` on the C++ side (GLlong doesn't exist and IIRC 64bit integers don't exist for OpenGL).

Right. Forgot.

Well then this is alr.

**Review by vaxerski — APPROVED** [review]
lgtm. thanks!

**Pull request merged**